impl/sql/mutationstorage/mutation_logs.go

```diff
@@ -173,21 +176,25 @@ func (m *Mutations) HighWatermark(ctx context.Context, directoryID string, logID
 			ORDER BY Q.Time ASC
 			LIMIT ?
 		) AS T1`,
-		directoryID, logID, start, batchSize).
+		directoryID, logID, startQuery, batchSize).
```

Comment:
I'm surprized the SQL query takes `time.Time` as a parameter. My understanding was that it was a microseconds int.

Have you considered not using the `DATETIME` type? There is much [details](https://dev.mysql.com/doc/refman/8.0/en/datetime.html) on how it works, which, together with details on how `time.Time` works, make me nervous about missing some unexpected corner cases like leap years/seconds or timezones, or keeping in mind this implicit "microsecond quantum" quirk.

Is MySQL `DATETIME` type one of the reasons you decided to use `time.Time`? I wonder if we could avoid this complexity as follows:

- Use `BIGINT` MySQL type for clarity
- Use `time.Now().UnixNano()` timestamp when writing
- Do not convert it back to `time.Time` when reading, just continue using the int
- Use `Watermark` type to wrap this int in a type-safe manner

Reply:
The big motivation for all these changes was managing the discrepancy between the mysql implementation (which used to be BIGINT with UnixNanos), and Spanner, which uses microseconds. These were two very different kinds of "ints", which caused their own problems.  Trying to abstract away the time precision of the storage layer is how we got to time.Time. Migrating to time.Time has also produced a whole new suite of tests to ensure that storage implementations implement these APIs correctly, and have removed quite a few sneaky modifications of these watermarks in the code.

I think your concerns are very legitimate and worth discussing further.